hydra-task/src/main/java/com/addthis/hydra/task/map/MapFeederForkJoin.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.addthis.hydra.task.map;
+
+
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.RecursiveAction;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.addthis.basis.util.Parameter;
+
+import com.addthis.bundle.core.Bundle;
+import com.addthis.hydra.task.source.TaskDataSource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class MapFeederForkJoin implements Runnable {
+    private static final Logger log = LoggerFactory.getLogger(MapFeederForkJoin.class);
+
+    private static final int QUEUE_DEPTH = Parameter.intValue("task.queue.depth", 100);
+
+    // state control
+    private final AtomicBoolean errored = new AtomicBoolean(false);
+    private boolean hasClosedStreams = false; // not shared with MapperTasks
+
+    // enclosing task
+    private final StreamMapper task;
+    private final TaskDataSource source;
+
+    // mapper task controls
+    private final ForkJoinPool mapperPool;
+    private final int parallelism;
+    private final Semaphore enqueuePermits;
+
+    private long start = System.currentTimeMillis();
+    private long totalBundles = 0;
+
+    public MapFeederForkJoin(StreamMapper task, TaskDataSource source, int parallelism) {
+        this.source = source;
+        this.task = task;
+        this.parallelism = parallelism;
+        this.mapperPool = new ForkJoinPool(parallelism);
```

Comment:
you can hand the constructor an uncaught exception handler to save on a lot of try/finally